### PR TITLE
[REVIEW] Add support for numpy/cupy array in `DataFrame` construction

### PR DESCRIPTION
**Pull request by galipremsagar**
Fixes: #5274 
Fixes: #5304 
Fixes: #5313

1. This PR adds support for arrays to be passed into the `DataFrame` constructor for which we are utilizing `from_records` API with some added functionality in case of a 1-d array input to it.
2. This PR also fixes issue with non-contiguous cuda array inputs where we didn't check if the array is contiguous or not before constructing the buffer out of the address pointer in `__cuda_array_interface__`.
3. This PR also fixes issues with array-like inputs for `index` in `DataFrame.from_records`.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -217,6 +217,13 @@ def __init__(self, data=None, index=None, columns=None, dtype=None):
                         ),
                     )
                 )
+        elif hasattr(data, "__array_interface__") or hasattr(
+            data, "__cuda_array_interface__"
+        ):
+            new_df = self.from_records(data, index=index, columns=columns)
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:223]: Can you update the docstring on `from_records` to make it clear it accepts array-like objects in either GPU or host memory?

Comment (kkraus14) [reply python/cudf/cudf/core/dataframe.py:223]: Actually, nevermind this isn't the right function that a general array should go through: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.from_records.html

Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py]: Looks like `Pandas.DataFrame.from_records` isn't accepting 1-d arrays, instead passing them through `_from_columns` flow.

File: python/cudf/cudf/core/column/column.py [hunk 2]
```diff
@@ -1229,6 +1229,11 @@ def as_column(arbitrary, nan_as_null=None, dtype=None, length=None):
             arbitrary = cupy.asarray(arbitrary).astype(arb_dtype)
             current_dtype = arb_dtype
 
+        if desc["strides"] is not None and desc.get("mask", None) is None:
```
Comment (jakirkham) [inline python/cudf/cudf/core/column/column.py:1232]: While `strides is None` is certainly one case that works, we should also check to see if `strides == (arb_dtype.itemsize,)` as that is also 1-D contiguous.

File: python/cudf/cudf/core/dataframe.py [hunk 3]
```diff
@@ -217,6 +217,31 @@ def __init__(self, data=None, index=None, columns=None, dtype=None):
...
+            arr_interface = data.__array_interface__
+            if (
+                len(arr_interface["shape"]) == 1
+                and len(arr_interface["descr"]) == 1
+            ):
+                new_df = self._from_columns(
+                    [data], index=index, columns=columns
+                )
+            else:
+                new_df = self.from_records(data, index=index, columns=columns)
```
Comment (jakirkham) [inline python/cudf/cudf/core/dataframe.py:230]: What happens if `len(arr_interface["shape"]) > 2`?

**Comment by jakirkham** [comment]
Thanks for working on this! 😄 Had a couple questions above

File: python/cudf/cudf/core/column/column.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py:1232]: Did you mean mean `if strides != (arb_dtype.itemsize,)` to check for 1-d non-contiguous?

I was not able to get a 1-d contiguous or non-contiguous array with such a condition, could you share an example?

```python
>>> cp.random.randn(10, 2)
array([[ 0.94513535,  0.83474103],
       [-0.48733383, -1.26725109],
       [ 0.92763362, -0.90840477],
       [ 1.51202585, -0.08442038],
       [-1.03553928,  2.14929072],
       [-0.64543545, -0.45847234],
       [-2.08376522, -0.82124291],
       [-0.72985977, -2.16106965],
       [ 0.88894551,  1.12507506],
       [-0.47210113, -1.03509318]])
>>> cp.random.randn(10, 2)[:, 0]
array([-1.40171207, -0.18492909, -0.67132592, -2.40527545,  1.49884205,
        2.01873371,  1.656129  , -1.00027593, -1.51870164,  0.17271589])
>>> cp.random.randn(10, 2)[:, 0].__cuda_array_interface__
{'shape': (10,), 'typestr': '<f8', 'descr': [('', '<f8')], 'version': 2, 'strides': (16,), 'data': (140367684960256, False)}
>>> np.dtype(cp.random.randn(10, 2)[:, 0].__cuda_array_interface__['typestr'])
dtype('float64')
>>> np.dtype(cp.random.randn(10, 2)[:, 0].__cuda_array_interface__['typestr']).itemsize
8
>>> cp.array([1, 2, 3])
array([1, 2, 3])
>>> cp.array([1, 2, 3]).__cuda_array_interface__
{'shape': (3,), 'typestr': '<i8', 'descr': [('', '<i8')], 'version': 2, 'strides': None, 'data': (140367684960256, False)}
>>> np.dtype(cp.array([1, 2, 3]).__cuda_array_interface__['typestr'])
dtype('int64')
>>> np.dtype(cp.array([1, 2, 3]).__cuda_array_interface__['typestr']).itemsize
8
```

File: python/cudf/cudf/core/dataframe.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py:230]: Which is a 2-d case, and all 2-d cases go through `from_records` (`from_records`) works with 2-d arrays too.

```python
>>> pd.DataFrame.from_records(np.random.randn(10, 2))
          0         1
0 -0.957835 -2.945332
1 -0.651440 -1.191708
2  0.419436  0.906071
3 -1.554972 -0.163832
4 -0.708032 -0.156609
5 -0.592104  0.664202
6  2.507729  0.780785
7 -0.330201 -1.405794
8 -0.734475 -1.065689
9 -0.639492  0.888378
```

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (kkraus14) [reply python/cudf/cudf/core/dataframe.py]: I don't think `from_records` is the intended path for 2-d arrays either. It's specifically meant for structured / record arrays.

Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py]: Not using `from_records` currently and introduced a new function `_from_arrays` to handle array-like inputs to dataframe.

File: python/cudf/cudf/core/column/column.py [hunk 4]
```diff
@@ -1207,6 +1207,11 @@ def as_column(arbitrary, nan_as_null=None, dtype=None, length=None):
             arbitrary = cupy.asarray(arbitrary).astype(arb_dtype)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/column.py:1207]: If arbitrary is masked this won't work.

File: python/cudf/cudf/core/column/column.py (hunk 2 above)
Comment (kkraus14) [reply python/cudf/cudf/core/column/column.py:1210]: +1 to supporting where `strides == dtype.itemsize`

File: python/cudf/cudf/core/dataframe.py [hunk 5]
```diff
@@ -217,6 +217,25 @@ def __init__(self, data=None, index=None, columns=None, dtype=None):
...
+            self.columns = new_df.columns
+        elif hasattr(data, "__cuda_array_interface__"):
+            arr_interface = data.__cuda_array_interface__
+            if len(arr_interface["descr"]) == 1:
+                new_df = self._from_arrays(data, index=index, columns=columns)
+            else:
+                new_df = self.from_records(data, index=index, columns=columns)
+            self._data = new_df._data
+            self._index = new_df._index
+            self.columns = new_df.columns
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:238]: Move this above the `__array_interface__` block to short circuit that check in case of a GPU object.

File: python/cudf/cudf/core/dataframe.py [hunk 6]
```diff
@@ -4265,8 +4286,64 @@ def from_records(self, data, index=None, columns=None, nan_as_null=False):
                 df[k] = Series(data[k], nan_as_null=nan_as_null)
 
         if index is not None:
-            indices = data[index]
-            return df.set_index(indices.astype(np.int64))
+            if is_scalar(index):
+                df = df.set_index(index)
+            else:
+                df._index = as_index(index)
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:4292]: Can we just go through the `df.index` setter here instead of using the internal attribute?

File: python/cudf/cudf/core/dataframe.py [hunk 7]
```diff
@@ -4265,8 +4286,64 @@ def from_records(self, data, index=None, columns=None, nan_as_null=False):
...
+        -------
+        DataFrame
+        """
+
+        data = cupy.asarray(data)
+        if data.ndim != 1 and data.ndim != 2:
+            raise ValueError(
+                "records dimension expected 1 or 2 but found {!r}".format(
+                    data.ndim
+                )
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:4320]: Use f-string instead of str.format

File: python/cudf/cudf/core/dataframe.py [hunk 8]
```diff
@@ -4265,8 +4286,64 @@ def from_records(self, data, index=None, columns=None, nan_as_null=False):
...
+            # Since we validate ndim to be either 1 or 2 above,
+            # this case can be assumed to be ndim == 1.
+            num_cols = 1
+
+        if columns is None:
+            names = [i for i in range(num_cols)]
+        else:
+            if len(columns) != num_cols:
+                msg = "columns length expected {!r} but found {!r}"
+                raise ValueError(msg.format(num_cols, len(columns)))
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:4335]: Use f-string

File: python/cudf/cudf/core/dataframe.py [hunk 9]
```diff
@@ -4265,8 +4286,64 @@ def from_records(self, data, index=None, columns=None, nan_as_null=False):
...
+            names = columns
+
+        df = DataFrame()
+        if data.ndim == 2:
+            for i, k in enumerate(names):
+                df[k] = Series(data[:, i], nan_as_null=nan_as_null)
+        elif data.ndim == 1:
+            df[names[0]] = Series(data, nan_as_null=nan_as_null)
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:4343]: Can we construct using columns to avoid unnecessary index alignment checking / handling?

File: python/cudf/cudf/core/dataframe.py [hunk 10]
```diff
@@ -4265,8 +4286,64 @@ def from_records(self, data, index=None, columns=None, nan_as_null=False):
...
+
+        if index is not None:
+            df._index = as_index(index)
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:4346]: Can we go through the `df.index` setter here?

File: python/cudf/cudf/core/dataframe.py [hunk 11]
```diff
@@ -4265,8 +4286,64 @@ def from_records(self, data, index=None, columns=None, nan_as_null=False):
                 df[k] = Series(data[k], nan_as_null=nan_as_null)
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:4286]: Shouldn't go through Series here either as it will do index alignment checking.

File: python/cudf/cudf/core/dataframe.py [hunk 12]
```diff
@@ -4335,7 +4412,7 @@ def to_gpu_matrix(self):
         )
         return self.as_gpu_matrix()
 
-    def _from_columns(cols, index=None, columns=None):
+    def _from_columns(self, cols, index=None, columns=None):
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:4415]: Why this change? Should this be a class method instead?

**Comment by kkraus14** [comment]
We may also want to add a deprecation warning to `from_gpu_matrix` / `from_matrix` with this change.

File: python/cudf/cudf/core/dataframe.py (hunk 6 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py:4292]: I think we'll specifically have to set internal attribute here since we have a check for previous and new_length which will fail in dataframe initialization cases.

https://github.com/rapidsai/cudf/blob/branch-0.15/python/cudf/cudf/core/dataframe.py#L2240-#L2242

Probably this is why rest of `DataFrame` initialization code aswell is setting index to `_index`.

File: python/cudf/cudf/core/column/column.py (hunk 2 above)
Comment (jakirkham) [reply python/cudf/cudf/core/column/column.py:1210]: Sure, how about this?

```python
In [1]: import cupy                                                             

In [2]: a = cupy.arange(5)                                                      

In [3]: a.itemsize, a.strides                                                   
Out[3]: (8, (8,))
```

File: python/cudf/cudf/core/dataframe.py (hunk 3 above)
Comment (jakirkham) [reply python/cudf/cudf/core/dataframe.py:226]: Sorry what I meant was what if it is 3-D or greater? Do we raise somewhere?

Comment (galipremsagar) [reply python/cudf/cudf/core/dataframe.py:225]: Yes we raise an error:

```python
>>> import cupy as cp
>>> a = cp.zeros((2, 3, 4))
>>> a
array([[[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]])
>>> a.ndim
3
>>> import cudf
>>> cudf.DataFrame(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/cudf/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/core/dataframe.py", line 223, in __init__
    new_df = self._from_arrays(data, index=index, columns=columns)
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/core/dataframe.py", line 4314, in _from_arrays
    f"records dimension expected 1 or 2 but found: {data.ndim}"
ValueError: records dimension expected 1 or 2 but found: 3

File: python/cudf/cudf/core/column/column.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py]: Thanks @jakirkham, this is now being handled.

File: python/cudf/cudf/core/column/column.py (hunk 4 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/column.py]: Resolved this issue by separating out mask before we create a cupy array.

**Comment by jakirkham** [comment]
Thanks Prem! 😄

**Comment by galipremsagar** [comment]
> One minor doc improvement, otherwise

I have another PR related to `DataFrame.__init__` in which I'll add some deprecated decorators which will show up in sphinx.

**Comment by kkraus14** [comment]
> > One minor doc improvement, otherwise
> 
> I have another PR related to `DataFrame.__init__` in which I'll add some deprecated decorators which will show up in sphinx.

That was a typo from a previously incorrect statement, so disregard that 😅
